draft-ietf-jsonpath-base.md

```diff
@@ -413,51 +433,70 @@ the grammar in this document, its UTF-8 form SHOULD first be decoded into
 Unicode code points as described
 in {{RFC3629}}.
 
-
-## Terminology
-
-A JSON value is logically a tree of nodes.
-
-Each node holds a JSON value (as defined by {{RFC8259}}) of one of the
-types object, array, number, string, or one of the literals `true`,
-`false`, or `null`.
-The type of the JSON value held by a node is
-sometimes referred to as the type of the node.
-
-
-## Implementation
-
-An implementation of this specification, from now on referred to simply as
-"an implementation", SHOULD takes two inputs, a JSONPath and a JSON value,
-and produce
-a possibly empty list of nodes of the JSON value which are selected by
-the JSONPath or an error (but not both).
-
-If no node is selected and no error has occurred, an implementation MUST
-return an empty list of nodes.
-
-Syntax errors in the JSONPath SHOULD be detected before selection is attempted
-since these errors do not depend on the JSON value.
-Therefore, an implementation SHOULD take a JSONPath and produce an optional
-syntax error and then,
-if and only if an error was not produced, SHOULD take a JSON value and
-produce a list of nodes or an error (but not both).
-
-Alternatively, an implementation MAY take a JSONPath and a JSON value
-and produce a list of nodes or an optional error (but not both).
-
-For any implementation, if a syntactically invalid JSONPath is provided,
-the implementation MUST return an error.
-
-If a syntactially invalid JSON value is provided, any implementation SHOULD
-return an error.
+A string to be used as a JSONPath query needs to be *well-formed* and
+*valid*.
+A string is a well-formed JSONPath query if it conforms to the syntax
+of JSONPath.
+A well-formed JSONPath query is valid if it also fulfills all semantic
+requirements posed by this document.
+
+Well-formedness and validity of JSONPath queries are independent of
+the JSON value the query is applied to; no further errors can be
+raised during application of the query to a JSON value.
+
+(Obviously, an implementation can still fail when executing a JSONPath
+query, e.g., because of resource depletion, but this is not modeled in
+the present specification.)
+
+> Discussion: are we ready for the error model defined here, i.e. one
+> that does not have runtime errors, but only compile time errors?
+
+## Processing Model
```

Comment:
We should add a paragraph that state that this processing model is an example for the purposes of explaining how JSON Path works and is not prescriptive of the internal workings of an implementation.  If an implementation wishes to (or needs to due to framework limitations) design a different process that yields the same results, this should be acceptable.

Reply:
Good point.  I put a new paragraph at the start of 3.2.  I had to weasel word a bit around "the same", because we are saying the model isn't deterministic.

Reply:
(Marking this as resolved -- it is however best if this is done by who raised the comment.)

Reply:
I looked for it last night, but it seems I can't mark it as resolved.